[flang] Fix the incorrect insertion point for alloca

erjin opened the pull request:
While creating a temporary alloca for a box in OpenMp region, the insertion point should be the OpenMP region block instead of the function entry block.

erjin commented:
This patch fixes the spec2017 cam4_s runtime error mentioned in https://github.com/llvm/llvm-project/issues/63147

jeanPerier commented on flang/lib/Optimizer/CodeGen/CodeGen.cpp:
```diff
@@ -333,25 +333,30 @@ class FIROpConversion : public mlir::ConvertOpToLLVMPattern<FromOp> {
+  // 2. A LLVMFuncOp
+  // 3. The first ancestor that is an OpenMP Op or a LLVMFuncOp
+  static mlir::Block *getBlockForAllocaInsert(mlir::Operation *op) {
+    if (auto iface =
+            mlir::dyn_cast<mlir::omp::OutlineableOpenMPOpInterface>(op)) {
+      return iface.getAllocaBlock();
+    } else if (auto llvmFuncOp = mlir::dyn_cast<mlir::LLVM::LLVMFuncOp>(op)) {
+      return &llvmFuncOp.front();
+    } else {
+      return getBlockForAllocaInsert(op->getParentOp());
+    }
```
small nit about the style flang lowering/codegen uses LLVM coding conventions:
- [no else after return](https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return)
- [no braces for single line if-else](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements)

erjin replied:
Hi @jeanPerier, thanks for the review! Modified based on your suggestions.

jeanPerier reviewed (commented):
Nit about the coding style, looks good otherwise, thanks.

kiranchandramohan reviewed (approved):
Thanks @erjin.

LG after addressing Jean's comment.

erjin committed:
[flang] Fix the incorrect insertion point for alloca

While creating a temporary alloca for a box in OpenMp region, the
insertion point should be the OpenMP region block instead of the
function entry block.

erjin commented on flang/lib/Optimizer/CodeGen/CodeGen.cpp:
```diff
@@ -333,25 +333,30 @@ class FIROpConversion : public mlir::ConvertOpToLLVMPattern<FromOp> {
-               : parentOp->getParentOfType<mlir::LLVM::LLVMFuncOp>();
+  // Find the Block in which the alloca should be inserted.
+  // The order to recursively find the proper block:
+  // 1. An OpenMP Op
```
Thanks! Modified.

jeanPerier reviewed (approved):
Thanks, LGTM

kiranchandramohan commented:
Thanks. Could you merge today itself and close the cam4_s ticket?

agozillon commented:
Thank you very much for this fix! I ran into this problem recently when trying to map a Fortran pointer, it's good to have a fix upstreamed.
